vowpalwabbit/core/src/reductions/interaction_ground.cc

```diff
@@ -3,111 +3,240 @@
-  double total_importance_weighted_reward = 0.;
-  double total_uniform_reward = 0.;
-  // the accumulated importance weighted loss of the policy which optimizes the negative of the given value
-  double total_importance_weighted_cost = 0.;
-  double total_uniform_cost = 0.;
+  VW::LEARNER::single_learner* ik_learner = nullptr;
+  VW::example* ik_ex = nullptr;
+
+  std::vector<std::vector<namespace_index>> ik_interactions;
+  std::vector<std::vector<extent_term>>* extent_interactions;
+
+  std::unique_ptr<VW::workspace> ik_all;
+  ftrl* ik_ftrl; // automatically save resume
+  std::unique_ptr<ftrl> pi_ftrl; // TODO: add save resume
+  ~interaction_ground() {
+    VW::dealloc_examples(ik_ex, 1);
+  }
+
+  // TODO: rule of 5
 };
 
-void negate_cost(VW::multi_ex& ec_seq)
+void learn(interaction_ground& igl, multi_learner& base, VW::multi_ex& ec_seq)
 {
-  for (auto* example_ptr : ec_seq)
+  float p_unlabeled_prior = 0.5f;
+
+  std::swap(igl.ik_ftrl->all->loss, igl.ik_all->loss);
+  std::swap(igl.ik_ftrl->all->sd, igl.ik_all->sd);
+
+  float ik_pred = 0.f;
+  int chosen_action_idx = 0;
+
+  const auto it = std::find_if(ec_seq.begin(), ec_seq.end(), [](VW::example* item) { return !item->l.cb.costs.empty(); });
+  if (it != ec_seq.end())
   {
-    for (auto& label : example_ptr->l.cb.costs)
-    {
-      if (label.has_observed_cost()) { label.cost = -label.cost; }
+    chosen_action_idx = std::distance(ec_seq.begin(), it);
+  }
+
+  auto feedback_ex = ec_seq.back(); // TODO: refactor. Now assume last example is feedback example
+  ec_seq.pop_back();
+
+  auto ns_iter = feedback_ex->indices.begin();
+  uint64_t feature_hash = *feedback_ex->feature_space.at(*ns_iter).indices.data();
+
+  for (auto& ex_action : ec_seq) {
+    VW::empty_example(*igl.ik_all, *igl.ik_ex);
+    // TODO: Do we need constant feature here? If so, VW::add_constant_feature
+    VW::details::append_example_namespaces_from_example(*igl.ik_ex, *ex_action);
+    igl.ik_ex->indices.push_back(VW::details::IGL_FEEDBACK_NAMESPACE);
+    igl.ik_ex->feature_space[VW::details::IGL_FEEDBACK_NAMESPACE].push_back(1, feature_hash); // TODO: is the feature value always 1?
+    igl.ik_ex->num_features++;
+
+    // 1. set up ik ex
+    float ik_label = -1.f;
+    if (!ex_action->l.cb.costs.empty()) {
+      ik_label = 1.f;
+    }
+    igl.ik_ex->l.simple.label = ik_label;
+
+    // TODO: update the importance for each example
+    float importance = 0.6f;
+    igl.ik_ex->weight = importance;
+
+    igl.ik_ex->interactions = &igl.ik_interactions;
+    igl.ik_ex->extent_interactions = igl.extent_interactions; // TODO(low pri): not reuse igl.extent_interactions, need to add in feedback
+
+    // 2. ik learn
+    igl.ik_learner->learn(*igl.ik_ex, 0);
+
+    if (!ex_action->l.cb.costs.empty()) {
+      ik_pred = igl.ik_ex->pred.scalar;
     }
+    VW::details::output_and_account_example(*igl.ik_ftrl->all, *igl.ik_ex);
```

Comment:
This cannot be done in the learn call. This operation is now split into `update_stats`, `output_example_prediction` and `print_update`